apps/studio/src/server/modules/user/inactiveUsers.service.ts

```diff
@@ -0,0 +1,241 @@
+          .where("Site.id", "in", siteIds)
+          .where("ResourcePermission.userId", "!=", userId) // don't want to ask users to ask themselves for permissions
+          .where("ResourcePermission.deletedAt", "is", null)
+          .where("ResourcePermission.role", "=", RoleType.Admin) // should only give the admin emails to request reactivation permissions from
+          .where("User.email", "not in", ISOMER_ADMINS_AND_MIGRATORS_EMAILS) // we don't want to send emails to admins and migrators
+          .select([
+            "Site.id as siteId",
+            db.fn.agg<string[]>("array_agg", ["User.email"]).as("adminEmails"),
+          ])
+          .groupBy("Site.id"),
+      )
+      // Needed as we still want the site records even if there are no other users with permissions for that site
+      .with("baseSites", (eb) =>
+        eb
+          .selectFrom("Site")
+          .where("Site.id", "in", siteIds)
+          .select(["Site.id as siteId", "Site.name as siteName"]),
+      )
+      .selectFrom("baseSites")
+      .leftJoin("siteAdmins", "siteAdmins.siteId", "baseSites.siteId")
+      .select([
+        "baseSites.siteName",
+        sql<string[]>`COALESCE("siteAdmins"."adminEmails", ARRAY[]::text[])`.as(
+          "adminEmails",
+        ),
+      ])
+      .execute()
+  )
+}
+
+type SendAccountDeactivationEmailsProps = AccountDeactivationEmailTemplateData[]
+const sendAccountDeactivationEmails = async (
+  deactivatedUsersAndSites: SendAccountDeactivationEmailsProps,
+): Promise<void> => {
+  for (const { recipientEmail, sitesAndAdmins } of deactivatedUsersAndSites) {
+    try {
+      await sendAccountDeactivationEmail({ recipientEmail, sitesAndAdmins })
+    } catch {
+      logger.error(
+        `Error sending account deactivation email for user ${recipientEmail}`,
+      )
+    }
+  }
+}
+
+export const bulkDeactivateInactiveUsers = async (): Promise<void> => {
+  // not setting fromDaysAgo to MAX_DAYS_FROM_LAST_LOGIN because we want to
+  // deactivate users beyond the MAX_DAYS_FROM_LAST_LOGIN (e.g. 91 days). This could be due to:
+  // 1. legacy users who wasn't covered by the MAX_DAYS_FROM_LAST_LOGIN before the feature was introduced
+  // 2. users who weren't removed for whatever reason e.g. unintentional failed cron job that wasn't retried
+  const inactiveUsers = await getInactiveUsers({})
+
+  const deactivatedUsersAndSites: SendAccountDeactivationEmailsProps = []
+
+  try {
+    await db
+      .transaction()
+      .setIsolationLevel("serializable")
+      .execute(async (tx) => {
+        const deactivatedUsersAndSiteIds = await deactivateUsers({
+          tx,
+          userIds: inactiveUsers.map((user) => user.id),
+        })
+
+        for (const { user, siteIds } of deactivatedUsersAndSiteIds) {
+          if (siteIds.length === 0) continue
+
+          const sitesAndAdmins = await getSiteAndAdmins({
+            tx,
+            userId: user.id,
+            siteIds,
+          })
```

Comment:
[question] we are still in the current `tx` so will the user who's being deactivated get an email also? not a big issue, i'm fine w leaving this in too lol

Reply:
good catch, updated in https://github.com/opengovsg/isomer/pull/1489/commits/e12c45f3cb63da2289d834c3dd46e727b6ad106a ! thanks!